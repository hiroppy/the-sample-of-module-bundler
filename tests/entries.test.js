'use strict';

const { readFile, mkdir } = require('fs').promises;
const { Script, runInContext, createContext } = require('vm');
const { join } = require('path');
const bundler = require('../src');
const { exportAllDeclaration } = require('@babel/types');

const fixtureBasePath = join(__dirname, 'fixtures');
const outputBasePath = join(__dirname, 'output');

async function build(fixturePath, outputPath, type) {
  await bundler({
    entry: join(fixturePath, type, 'entry.js'),
    output: join(outputPath, type, 'main.js'),
  });
}

async function runGeneratedCodeInVM(outputPath, lastType) {
  const code = await readFile(join(outputPath, lastType, 'main.js'), 'utf-8');
  const sandbox = { console, process };
  const ctx = new createContext(sandbox);

  runInContext(code, ctx);
}

async function createDir(base, lastType) {
  await mkdir(join(base, lastType), { recursive: true });
}

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation();
  jest.spyOn(console, 'warn').mockImplementation();
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('common', () => {
  const fixturePath = join(fixtureBasePath, 'common');
  const outputPath = join(outputBasePath, 'common');

  beforeAll(async () => {
    await createDir(outputPath, '');
  });

  test('not found a module', async () => {
    try {
      await createDir(outputPath, 'notFoundModule');
      await build(fixturePath, outputPath, 'notFoundModule');
    } catch (e) {
      expect(e.message.includes('could not find the module:')).toBeTruthy();
    }
  });

  test('interop', async () => {
    await createDir(outputPath, 'interop');
    await build(fixturePath, outputPath, 'interop');
    await runGeneratedCodeInVM(outputPath, 'interop');

    expect(console.log.mock.calls).toMatchSnapshot();
  });
});

describe('cjs', () => {
  const fixturePath = join(fixtureBasePath, 'cjs');
  const outputPath = join(outputBasePath, 'cjs');

  beforeAll(async () => {
    await createDir(outputPath, '');
  });

  const dirs = ['simple', 'nested', 'circularDependency', 'filename', 'node-modules'];

  for (const dir of dirs) {
    test(dir, async () => {
      await createDir(outputPath, dir);
      await build(fixturePath, outputPath, dir);
      await runGeneratedCodeInVM(outputPath, dir);

      expect(console.log.mock.calls).toMatchSnapshot();
    });
  }
});

describe('esm', () => {
  const fixturePath = join(fixtureBasePath, 'esm');
  const outputPath = join(outputBasePath, 'esm');

  beforeAll(async () => {
    await createDir(outputPath, '');
  });

  const dirs = [
    'simple',
    'complication',
    'node-modules',
    'default',
    'specifier',
    'namespaceSpecifier',
    'import',
  ];

  for (const dir of dirs) {
    test(dir, async () => {
      await createDir(outputPath, dir);
      await build(fixturePath, outputPath, dir);
      await runGeneratedCodeInVM(outputPath, dir);

      expect(console.log.mock.calls).toMatchSnapshot();
    });
  }
});
