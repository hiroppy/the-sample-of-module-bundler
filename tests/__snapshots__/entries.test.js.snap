// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cjs circularDependency 1`] = `
Array [
  Array [
    "module1:",
    "from entry",
  ],
  Array [
    "main:",
    "from module1",
  ],
]
`;

exports[`cjs filename 1`] = `
Array [
  Array [
    "entry",
  ],
  Array [
    "module1",
  ],
  Array [
    "module2",
  ],
  Array [
    "index",
  ],
]
`;

exports[`cjs nested 1`] = `
Array [
  Array [
    "module1",
  ],
  Array [
    "module1-sub",
  ],
  Array [
    "entry",
  ],
]
`;

exports[`cjs node-modules 1`] = `
Array [
  Array [
    "16.13.1",
  ],
]
`;

exports[`cjs simple 1`] = `
Array [
  Array [
    "entry",
  ],
  Array [
    "module1",
  ],
  Array [
    "module2: value",
  ],
  Array [
    "module2: fn",
  ],
  Array [
    "module3",
    "from module1.js",
  ],
  Array [
    "sep----",
  ],
  Array [
    "module3",
    "from entry.js",
  ],
]
`;

exports[`common interop 1`] = `
Array [
  Array [
    "---- default.js ----",
  ],
  Array [
    "default from esm",
  ],
  Array [
    "default from cjs",
  ],
  Array [
    "---- default.js ----",
  ],
  Array [
    "hello world from cjs",
  ],
  Array [
    "hello world from esm",
  ],
  Array [
    "default from cjs",
  ],
  Array [
    "default from esm",
  ],
]
`;

exports[`esm complication 1`] = `
Array [
  Array [
    "1. default: named",
    "named: module2Value",
  ],
  Array [
    "2. default: arguments",
    "from entry-point",
  ],
  Array [
    "3. default: anonymous",
    "from module1 from entry-point",
  ],
  Array [
    "4. named: func",
  ],
  Array [
    6,
  ],
  Array [
    "5. module3 value",
  ],
  Array [
    "5. module3 value",
  ],
]
`;

exports[`esm default 1`] = `
Array [
  Array [
    "hi",
  ],
]
`;

exports[`esm import 1`] = `
Array [
  Array [
    "hello world",
  ],
]
`;

exports[`esm namespaceSpecifier 1`] = `
Array [
  Array [
    "hi",
    "world",
  ],
]
`;

exports[`esm node-modules 1`] = `
Array [
  Array [
    "16.13.1",
  ],
  Array [
    "16.13.1",
  ],
]
`;

exports[`esm simple 1`] = `
Array [
  Array [
    20,
  ],
]
`;

exports[`esm specifier 1`] = `
Array [
  Array [
    "hi",
    "world",
  ],
]
`;
